### ISSUE TITLE
[FSDP] Got error messages when model output is tuple

### ISSUE THREAD

**Issue by foreveronehundred** [facebookresearch/fairscale#967]
FSDP version: 1bc96fa8c69def6d990e42bfbd75f86146ce29bd
I wrapped my model following the same way as the pseudo code,
```
    # Wrap the model into FSDP, which will reduce parameters to the proper ranks
    model = FSDP(model, mixed_precision=True, reshard_after_forward = False)

    # Any relevant training loop. For example:
    model.train()
    for e in range(epochs):
        for (data, target) in dataloader:
            data, target = data.to(rank), target.to(rank)
            # Train
            model.zero_grad()
            outputs = model(data)
            loss = loss_fn(outputs, target)
            loss.backward()
            optimizer.step()
```
In my case, the outputs is a tuple (tensor1, tensor2, tensor3), and three tensors are used to calculate the loss. However, I found error messages when I call the loss.backward(),

```
  File "/home/XXX/.local/lib/python3.8/site-packages/Trainer.py", line 1456, in backward_pass
    loss.backward()
  File "/home/XXX/.local/lib/python3.8/site-packages/torch/_tensor.py", line 255, in backward
    torch.autograd.backward(self, gradient, retain_graph, create_graph, inputs=inputs)
  File "/home/XXX/.local/lib/python3.8/site-packages/torch/autograd/__init__.py", line 147, in backward
    Variable._execution_engine.run_backward(
  File "/home/XXX/.local/lib/python3.8/site-packages/fairscale/nn/data_parallel/fully_sharded_data_parallel.py", line 1495, in _pre_backward_hook
    self._use_full_params()
  File "/home/XXX/.local/lib/python3.8/site-packages/torch/autograd/grad_mode.py", line 28, in decorate_context
    return func(*args, **kwargs)
  File "/home/XXX/.local/lib/python3.8/site-packages/fairscale/nn/data_parallel/fully_sharded_data_parallel.py", line 2006, in _use_full_params
    assert self.has_full_params
```

By the way, there is no bugs when I use  FSDP autowrap, but I still want the try the above wrapping procedure since I think it can bring better performance.

**Comment by anj-s** [comment]
Is there just one outer wrap that is raising this error? With autowrap is the wrapping different? Can you share an output showing the difference?

**Comment by foreveronehundred** [comment]
> Is there just one outer wrap that is raising this error?
Yes.
> With autowrap is the wrapping different?
With autowrap, every leaf module is wraped
> Can you share an output showing the difference?
I think I can give an example
with autowrap:
```
model.blocks.layer1
model.blocks.layer1._fsdp_wrapped_module
model.blocks.layer1._fsdp_wrapped_module._fpw_module
model.blocks.layer2
model.blocks.layer2._fsdp_wrapped_module
model.blocks.layer2._fsdp_wrapped_module._fpw_module
model.blocks.layer3
model.blocks.layer3._fsdp_wrapped_module
model.blocks.layer3._fsdp_wrapped_module._fpw_module
```
without autowrap:
```
_fsdp_wrapped_module._fpw_module.model.blocks.layer1
_fsdp_wrapped_module._fpw_module.model.blocks.layer2
_fsdp_wrapped_module._fpw_module.model.blocks.layer3
```

**Comment by foreveronehundred** [comment]
Here is some details, and I hope it can help you fix the problems.

In my case, three tensors in outputs is registered with pre_backward_hooks() and post_backward_hooks(). According to my observation, the pre_backward_hooks() of tensor1 is first executed, then post_backward_hooks() of tensor1 is executed. Later one is the pre_backward_hooks() of tensor2, and so on.
However, self.has_full_params will be set to False in post_backward_hooks() of tensor1, and the assertion in pre_backward_hooks() of tensor2 will fail because of that.

**Comment by min-xu-ai** [comment]
there has a been a few fixes that might address this. Can you try the 0.4.10 release if you still care about this case? Thanks a lot!
